File: Connection/InformationPacket.cs
using System.Linq;
using System.Text.Json;
using System.Text;

namespace LANchat.Connection
{
    internal class InformationPacket : AbstractTcpPacket
    {
        public AppSettings Information;

        public InformationPacket(AppSettings info) : base(PacketDataType.INFO)
        {
            Information = info;
        }

        public InformationPacket(byte[] bytes) : base(PacketDataType.INFO)
        {
            Information = JsonSerializer.Deserialize<AppSettings>(bytes.Skip(1).ToArray(),
                new JsonSerializerOptions() { IncludeFields = true });
        }

        public override byte[] ToByteArray()
        {
            return Encoding.UTF8.GetBytes(JsonSerializer.Serialize(Information, new JsonSerializerOptions()
            { IncludeFields = true })).Prepend(base.ToByteArray()[0]).ToArray();
        }

        public override void Apply(MainWindow mw, ChatClient client)
        {
            mw.UpdateClientInfo(client, Information);
        }
    }
}

File: connection/informationpacket.cpp
#include "informationpacket.h"
#include <QJsonDocument>
#include <QJsonObject>

namespace Connection
{
    InformationPacket::InformationPacket(AppSettings info)
        : AbstractTcpPacket(PacketDataType::INFO),
          Information(info)
    { }

    InformationPacket::InformationPacket(QByteArray& bytes) : AbstractTcpPacket(PacketDataType::INFO)
    {
        QJsonDocument doc = QJsonDocument::fromJson(bytes);
        Information.Username = doc.object().value("Username").toString();
        Information.Port = doc.object().value("Port").toInt();
    }

    void InformationPacket::Apply(MainWindow &mw, ChatClient &client)
    {
        mw.UpdateClientInfo(client, Information);
    }

    QByteArray InformationPacket::ToByteArray()
    {
        QJsonDocument doc(QJsonObject{
            { "Username", Information.Username },
            { "Port", Information.Port }
        });
        return AbstractTcpPacket::ToByteArray().append(doc.toJson());
    }
} // namespace Connection
